Public/Src/Cache/ContentStore/Distributed/Utilities/GrpcFileCopier.cs

```diff
@@ -0,0 +1,92 @@
+    /// </summary>
+    public class GrpcFileCopier : IAbsolutePathFileCopier
+    {
+        private Context _context;
+        private int _grpcPort;
+
+        /// <summary>
+        /// Constructor for <see cref="GrpcFileCopier"/>.
+        /// </summary>
+        public GrpcFileCopier(Context context, int grpcPort)
+        {
+            _context = context;
+            _grpcPort = grpcPort;
+        }
+
+        /// <inheritdoc />
+        public Task<FileExistenceResult> CheckFileExistsAsync(AbsolutePath path, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            // TODO: Implement!
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc />
+        public async Task<CopyFileResult> CopyFileAsync(AbsolutePath sourcePath, AbsolutePath destinationPath, long contentSize, bool overwrite, CancellationToken cancellationToken)
+        {
+            // Extract host and contentHash from sourcePath
+            (string host, ContentHash contentHash) = ExtractHostHashFromAbsolutePath(sourcePath);
+
+            CopyFileResult copyFileResult = null;
+            // Contact hard-coded port on source
+            using (var client = GrpcCopyClient.Create(System.Net.IPAddress.Any.ToString(), 7089))
+            {
+                copyFileResult = await client.CopyFileAsync(_context, contentHash, destinationPath, contentSize);
+            }
+
+            return copyFileResult;
+        }
+
+        private (string host, ContentHash contentHash) ExtractHostHashFromAbsolutePath(AbsolutePath sourcePath)
+        {
+            Contract.Assert(sourcePath.IsUnc);
+
+            // TODO: Keep the segments in the AbsolutePath object?
+            // TODO: Indexable structure?
+            var segments = sourcePath.GetSegments();
+            Contract.Assert(segments.Count >= 4);
+
+            var host = segments.First();
+            var hashLiteral = segments.Last();
+            if (hashLiteral.EndsWith(GrpcDistributedPathTransformer.BlobFileExtension))
+            {
+                hashLiteral = hashLiteral.Substring(0, hashLiteral.Length - GrpcDistributedPathTransformer.BlobFileExtension.Length);
+            }
+            var hashTypeLiteral = segments.ElementAt(segments.Count - 1 - 2);
+
+            if (!Enum.TryParse<HashType>(hashTypeLiteral, out HashType hashType))
+            {
+                throw new InvalidOperationException($"{hashTypeLiteral} is not a valid member of {nameof(HashType)}");
+            }
+
+            var contentHash = new ContentHash(hashType, HexUtilities.HexToBytes(hashLiteral));
+
+            return (host, contentHash);
+        }
+
+        /// <inheritdoc />
+        public Task<CopyFileResult> CopyToAsync(AbsolutePath sourcePath, Stream destinationStream, long expectedContentSize, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();

```

Comment:
>throw [](start = 12, length = 5)

Will this screw with hash-while-streaming?

Reply:
Absolutely. This still needs to be done. If trusted hash is off, then this will not be called.

---
In reply to: [271087833](https://github.com/Microsoft/BuildXL/pull/66#discussion_r271087833) [](ancestors = 271087833)